pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+func (m *kubeGenericRuntimeManager) Version() (kubecontainer.Version, error) {
+	typedVersion, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Get remote runtime version failed: %v", err)
+		return nil, err
+	}
+
+	return newRuntimeVersion(typedVersion.GetVersion())
+}
+
+// APIVersion returns the cached API version information of the container
+// runtime. Implementation is expected to update this cache periodically.
+// This may be different from the runtime engine's version.
+func (m *kubeGenericRuntimeManager) APIVersion() (kubecontainer.Version, error) {
+	typedVersion, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Get remote runtime version failed: %v", err)
+		return nil, err
+	}
+
+	return newRuntimeVersion(typedVersion.GetRuntimeApiVersion())
+}
+
+// Status returns error if the runtime is unhealthy; nil otherwise.
+func (m *kubeGenericRuntimeManager) Status() error {
+	_, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Checkout remote runtime status failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// GetPods returns a list of containers grouped by pods. The boolean parameter
+// specifies whether the runtime returns all containers including those already
+// exited and dead containers (used for garbage collection).
+func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, error) {
+	pods := make(map[kubetypes.UID]*kubecontainer.Pod)
+	var result []*kubecontainer.Pod
+
+	sandboxes, err := m.getKubeletPodSandboxes(all)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range sandboxes {
+		name, namespace, _ := kubecontainer.ParsePodFullName(s.GetName())
+		labelledPodInfo := GetPodSandboxInfoFromLabels(s.Labels)
+		pods[labelledPodInfo.PodUID] = &kubecontainer.Pod{
+			ID:        labelledPodInfo.PodUID,
+			Name:      name,
+			Namespace: namespace,
+		}
+	}
+
+	containers, err := m.getKubeletContainers(all)
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range containers {
+		labelledInfo := GetContainerInfoFromLabels(c.Labels)
+		pod, found := pods[labelledInfo.PodUID]
+		if !found {
+			glog.Warningf("Can not find sandbox %v for container %s", labelledInfo.PodUID, c.Name)
+			continue
+		}
+
+		converted, err := m.toRuntimeContainer(c)
+		if err != nil {
+			glog.Warningf("Convert %s container %v failed: %v", m.runtimeType, c, err)
+			continue
+		}
+
+		pod.Containers = append(pod.Containers, converted)
+	}
+
+	// Convert map to list.
+	for _, p := range pods {
+		result = append(result, p)
+	}
+
+	return result, nil
+}
+
+func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod) (*runtimeApi.PodSandboxConfig, error) {
+	podFullName := kubecontainer.BuildPodFullName(pod.Name, pod.Namespace)
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &podFullName,
+		Labels:      newPodLabels(pod),
```

Comment:
Labels or annotations?
IIRC, annotation is for arbitrary data, label is for select.
To me, grace period or something looks more like annotation. :)